File: Library/Plugin/AopPlugin/src/main/java/org/alee/component/skin/page/WindowManager.java
package org.alee.component.skin.page;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;

import org.alee.component.skin.plugin.ISkinOptionFactoryLoader;
import org.alee.component.skin.service.ThemeSkinService;
import org.alee.component.skin.util.PrintUtil;

import java.util.WeakHashMap;

/**********************************************************
 *
 * @author: MY.Liu
 * @date: 2021/2/15
 * @description: xxxx
 *
 *********************************************************/
public final class WindowManager {

    private final WeakHashMap<Context, BaseWindowProxy> mWindowProxyWeakHashMap;

    private boolean mIsInitialized;

    private ISkinOptionFactoryLoader mOptionFactoryLoader;

    private WindowManager() {
        mWindowProxyWeakHashMap = new WeakHashMap<>();
        try {
            mOptionFactoryLoader = (ISkinOptionFactoryLoader) getClass().getClassLoader().loadClass("org.alee.component.skin.plugin.SkinOptionFactoryLoader").newInstance();
        } catch (Throwable e) {
            PrintUtil.getInstance().printE(e);
        }
    }

    /**
     * 获取单例对象
     *
     * @return {@link WindowManager}
     */
    public static WindowManager getInstance() {
        return WindowManagerHolder.INSTANCE;
    }

    void init(@NonNull Application application) {
        if (mIsInitialized) {
            return;
        }
        mIsInitialized = true;
        ThemeSkinService.getInstance().init(application.getApplicationContext(), mOptionFactoryLoader.load());
        application.registerActivityLifecycleCallbacks(new ActivityLifecycleCallbacks());
        onAttachWindow(application.getApplicationContext(), new GlobalWindowProxy());
    }


    private void onAttachWindow(@NonNull Context context, @NonNull BaseWindowProxy proxy) {
        synchronized (mWindowProxyWeakHashMap) {
            if (mWindowProxyWeakHashMap.containsKey(context)) {
                return;
            }
            proxy.bindLayoutInflate(context);
            mWindowProxyWeakHashMap.put(context, proxy);
        }
    }

    void onAttachWindow(@NonNull Context context) {
        onAttachWindow(context, new ActivityWindowProxy());
    }

    void onDetachWindow(Context context) {
        synchronized (mWindowProxyWeakHashMap) {
            BaseWindowProxy windowProxy = mWindowProxyWeakHashMap.remove(context);
            if (null != windowProxy) {
                windowProxy.onDestroy();
            }
        }
    }

    public IWindowProxy getWindowProxy(Context context) {
        return mWindowProxyWeakHashMap.get(context);
    }

    /**
     * 静态内部类持有外部对象实现单利方式
     */
    private static class WindowManagerHolder {
        /**
         * {@link WindowManager}
         */
        private static final WindowManager INSTANCE = new WindowManager();
    }
}

File: Library/PaintedSkin-Kotlin/src/main/java/org/alee/component/skin/core/window/WindowManager.kt
package org.alee.component.skin.core.window

import android.app.Activity
import android.app.Application
import android.content.Context
import java.util.WeakHashMap

/**
 * 摘要
 *
 * <p> 详细描述
 *
 * @author MingYu.Liu
 * created in 2023/9/12
 *
 */
internal object WindowManager {
    /**
     * Activity 生命周期观察者
     */
    private val mActivityLifecycleObserver by lazy { ActivityLifecycleObserver() }

    /**
     * 窗口代理集合
     */
    private val mWindowProxyMap: MutableMap<Context, IWindowProxy> by lazy { WeakHashMap() }

    internal fun init(application: Application) {
        application.registerActivityLifecycleCallbacks(mActivityLifecycleObserver)
        val globalWindow = GlobalWindowProxy()
        onWindowCreated(application, globalWindow)
        globalWindow.onWindowVisible()
    }

    internal fun onWindowCreated(activity: Activity) {
        onWindowCreated(activity, ActivityWindowProxy(activity.javaClass.simpleName))
    }

    internal fun onWindowDestroy(context: Context) {
        synchronized(mWindowProxyMap) {
            mWindowProxyMap.remove(context)?.run {
                onWindowInvisible()
                onWindowDestroyed()
            }
        }
    }

    internal fun fetchWindowProxy(context: Context): IWindowProxy? {
        return synchronized(mWindowProxyMap) { mWindowProxyMap[context] }
    }

    internal suspend fun dump(): String {
        return synchronized(mWindowProxyMap) {
            val sb = StringBuilder()
            mWindowProxyMap.values.forEach {
                sb.append(it.dump())
            }
            sb.toString()
        }
    }

    private fun onWindowCreated(context: Context, proxy: IWindowProxy) {
        synchronized(mWindowProxyMap) {
            if (mWindowProxyMap.containsKey(context)) {
                return
            }
            mWindowProxyMap[context] = proxy
            proxy.onWindowCreated(context)
        }
    }
}
